File: Countdown.cs
// Countdown.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

class Event
{
    public string Name { get; set; }
    public string Date { get; set; }
    public string Time { get; set; }
}

class Countdown
{
    private static List<Event> events = new List<Event>();
    private const string EventsFile = "events.json";

    static void LoadEvents()
    {
        try
        {
            string json = File.ReadAllText(EventsFile);
            events = JsonSerializer.Deserialize<List<Event>>(json) ?? new List<Event>();
        }
        catch { events = new List<Event>(); }
    }

    static void SaveEvents()
    {
        string json = JsonSerializer.Serialize(events, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(EventsFile, json);
    }

    static DateTime? ParseDate(string dateStr)
    {
        string[] formats = { "yyyy-MM-dd", "dd/MM/yyyy", "MM/dd/yyyy", "dd.MM.yyyy" };
        foreach (var fmt in formats)
        {
            if (DateTime.TryParseExact(dateStr, fmt, null, System.Globalization.DateTimeStyles.None, out DateTime result))
                return result;
        }
        return null;
    }

    static Event FindEvent(string name)
    {
        return events.Find(e => e.Name.Equals(name, StringComparison.OrdinalIgnoreCase));
    }

    static bool AddEvent(string name, string dateStr, string timeStr)
    {
        if (ParseDate(dateStr) == null) return false;
        if (string.IsNullOrEmpty(timeStr)) timeStr = "00:00";
        events.Add(new Event { Name = name, Date = dateStr, Time = timeStr });
        SaveEvents();
        return true;
    }

    static bool DeleteEvent(string name)
    {
        var ev = FindEvent(name);
        if (ev != null)
        {
            events.Remove(ev);
            SaveEvents();
            return true;
        }
        return false;
    }

    static string FormatCountdown(TimeSpan delta)
    {
        if (delta.TotalSeconds < 0) return "Event has passed!";
        int totalSec = (int)delta.TotalSeconds;
        int years = (int)(totalSec / (365.25 * 24 * 3600));
        totalSec -= (int)(years * 365.25 * 24 * 3600);
        int months = (int)(totalSec / (30.44 * 24 * 3600));
        totalSec -= (int)(months * 30.44 * 24 * 3600);
        int days = totalSec / (24 * 3600);
        totalSec -= days * (24 * 3600);
        int hours = totalSec / 3600;
        totalSec -= hours * 3600;
        int minutes = totalSec / 60;
        int seconds = totalSec % 60;
        var parts = new List<string>();
        if (years > 0) parts.Add($"{years} year{(years > 1 ? "s" : "")}");
        if (months > 0) parts.Add($"{months} month{(months > 1 ? "s" : "")}");
        if (days > 0) parts.Add($"{days} day{(days > 1 ? "s" : "")}");
        if (hours > 0) parts.Add($"{hours} hour{(hours > 1 ? "s" : "")}");
        if (minutes > 0) parts.Add($"{minutes} minute{(minutes > 1 ? "s" : "")}");
        if (seconds > 0) parts.Add($"{seconds} second{(seconds > 1 ? "s" : "")}");
        return parts.Count > 0 ? string.Join(", ", parts) : "0 seconds";
    }

    static void Countdown(Event ev, string formatStyle, bool live)
    {
        var timeParts = ev.Time.Split(':');
        int h = int.Parse(timeParts[0]);
        int m = int.Parse(timeParts[1]);
        var target = DateTime.Parse(ev.Date).AddHours(h).AddMinutes(m);
        var now = DateTime.Now;
        var delta = target - now;
        if (delta.TotalSeconds <= 0)
        {
            Console.WriteLine($"⏰ {ev.Name} is in the past!");
            return;
        }
        if (live)
        {
            while (delta.TotalSeconds > 0)
            {
                Console.Write($"\r⏳ {ev.Name}: {FormatCountdown(delta)}    ");
                System.Threading.Thread.Sleep(1000);
                now = DateTime.Now;
                delta = target - now;
            }
            Console.WriteLine($"\r🎉 {ev.Name} has arrived!    ");
        }
        else
        {
            if (formatStyle == "compact")
            {
                int totalSec = (int)delta.TotalSeconds;
                int days = totalSec / 86400;
                int hours = (totalSec % 86400) / 3600;
                int minutes = (totalSec % 3600) / 60;
                int seconds = totalSec % 60;
                Console.WriteLine($"⏳ {ev.Name}: {days:D2}:{hours:D2}:{minutes:D2}:{seconds:D2}");
            }
            else if (formatStyle == "short")
            {
                Console.WriteLine($"⏳ {ev.Name}: {(int)delta.TotalSeconds} seconds");
            }
            else
            {
                Console.WriteLine($"⏳ {ev.Name}: {FormatCountdown(delta)}");
            }
        }
    }

    static void ListEvents()
    {
        if (events.Count == 0)
        {
            Console.WriteLine("No events saved.");
            return;
        }
        foreach (var e in events)
        {
            var timeParts = e.Time.Split(':');
            int h = int.Parse(timeParts[0]);
            int m = int.Parse(timeParts[1]);
            var target = DateTime.Parse(e.Date).AddHours(h).AddMinutes(m);
            string status = target > DateTime.Now ? "⏳ pending" : "✅ passed";
            Console.WriteLine($"  {e.Name}: {e.Date} {e.Time} - {status}");
        }
    }

    static void Main(string[] args)
    {
        LoadEvents();
        if (args.Length > 0)
        {
            // CLI mode
            var dict = new Dictionary<string, string>();
            for (int i = 0; i < args.Length; i++)
            {
                if (args[i].StartsWith("--"))
                {
                    string key = args[i].Substring(2);
                    if (i + 1 < args.Length && !args[i + 1].StartsWith("--"))
                        dict[key] = args[++i];
                    else
                        dict[key] = "true";
                }
            }
            if (dict.ContainsKey("list")) { ListEvents(); return; }
            if (dict.ContainsKey("delete"))
            {
                if (DeleteEvent(dict["delete"])) Console.WriteLine("Event deleted.");
                else Console.WriteLine("Event not found.");
                return;
            }
            if (dict.ContainsKey("event") && dict.ContainsKey("date"))
            {
                var ev = FindEvent(dict["event"]);
                if (ev != null)
                {
                    string format = dict.ContainsKey("format") ? dict["format"] : "full";
                    bool live = dict.ContainsKey("live");
                    Countdown(ev, format, live);
                }
                else Console.WriteLine("Event not found.");
            }
            else Console.WriteLine("Usage: Countdown --event NAME --date DATE [--time TIME] [--format full|compact|short] [--live]");
            return;
        }

        // Interactive mode
        Console.WriteLine("=== Countdown Timer ===");
        while (true)
        {
            Console.WriteLine("\n1. Set new event");
            Console.WriteLine("2. Show countdown");
            Console.WriteLine("3. Live countdown");
            Console.WriteLine("4. List events");
            Console.WriteLine("5. Delete event");
            Console.WriteLine("6. Save events");
            Console.WriteLine("7. Load events");
            Console.WriteLine("8. Exit");
            Console.Write("Choose: ");
            string choice = Console.ReadLine()?.Trim();
            switch (choice)
            {
                case "1":
                    Console.Write("Event name: ");
                    string name = Console.ReadLine()?.Trim();
                    if (string.IsNullOrEmpty(name)) { Console.WriteLine("Name required."); break; }
                    if (FindEvent(name) != null) { Console.WriteLine("Event already exists."); break; }
                    Console.Write("Event date (YYYY-MM-DD): ");
                    string dateStr = Console.ReadLine()?.Trim();
                    if (ParseDate(dateStr) == null) { Console.WriteLine("Invalid date."); break; }
                    Console.Write("Event time (HH:MM, optional): ");
                    string timeStr = Console.ReadLine()?.Trim();
                    if (string.IsNullOrEmpty(timeStr)) timeStr = "00:00";
                    if (AddEvent(name, dateStr, timeStr)) Console.WriteLine("Event saved.");
                    break;
                case "2":
                    Console.Write("Event name: ");
                    name = Console.ReadLine()?.Trim();
                    var ev = FindEvent(name);
                    if (ev == null) { Console.WriteLine("Event not found."); break; }
                    Countdown(ev, "full", false);
                    break;
                case "3":
                    Console.Write("Event name: ");
                    name = Console.ReadLine()?.Trim();
                    ev = FindEvent(name);
                    if (ev == null) { Console.WriteLine("Event not found."); break; }
                    Countdown(ev, "full", true);
                    break;
                case "4": ListEvents(); break;
                case "5":
                    Console.Write("Event name to delete: ");
                    name = Console.ReadLine()?.Trim();
                    if (DeleteEvent(name)) Console.WriteLine("Event deleted.");
                    else Console.WriteLine("Event not found.");
                    break;
                case "6": SaveEvents(); Console.WriteLine("Events saved."); break;
                case "7": LoadEvents(); Console.WriteLine("Events loaded."); break;
                case "8": Console.WriteLine("Goodbye!"); return;
                default: Console.WriteLine("Invalid choice."); break;
            }
        }
    }
}

File: Countdown.java
// Countdown.java
import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.time.format.*;
import java.util.*;
import com.google.gson.*;

public class Countdown {
    private static List<Event> events = new ArrayList<>();
    private static final String EVENTS_FILE = "events.json";
    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();

    static class Event {
        String name;
        String date;
        String time;
    }

    static void loadEvents() {
        try {
            String json = new String(Files.readAllBytes(Paths.get(EVENTS_FILE)));
            Event[] arr = gson.fromJson(json, Event[].class);
            if (arr != null) events = new ArrayList<>(Arrays.asList(arr));
        } catch (Exception e) {
            events = new ArrayList<>();
        }
    }

    static void saveEvents() {
        String json = gson.toJson(events);
        try {
            Files.write(Paths.get(EVENTS_FILE), json.getBytes());
        } catch (Exception e) {}
    }

    static LocalDate parseDate(String dateStr) {
        DateTimeFormatter[] formatters = {
            DateTimeFormatter.ofPattern("yyyy-MM-dd"),
            DateTimeFormatter.ofPattern("dd/MM/yyyy"),
            DateTimeFormatter.ofPattern("MM/dd/yyyy"),
            DateTimeFormatter.ofPattern("dd.MM.yyyy")
        };
        for (DateTimeFormatter fmt : formatters) {
            try {
                return LocalDate.parse(dateStr, fmt);
            } catch (DateTimeParseException e) {}
        }
        return null;
    }

    static Event findEvent(String name) {
        for (Event e : events) {
            if (e.name.equalsIgnoreCase(name)) return e;
        }
        return null;
    }

    static boolean addEvent(String name, String dateStr, String timeStr) {
        if (parseDate(dateStr) == null) return false;
        if (timeStr == null || timeStr.isEmpty()) timeStr = "00:00";
        Event e = new Event();
        e.name = name;
        e.date = dateStr;
        e.time = timeStr;
        events.add(e);
        saveEvents();
        return true;
    }

    static boolean deleteEvent(String name) {
        for (int i = 0; i < events.size(); i++) {
            if (events.get(i).name.equalsIgnoreCase(name)) {
                events.remove(i);
                saveEvents();
                return true;
            }
        }
        return false;
    }

    static String formatCountdown(long seconds) {
        if (seconds < 0) return "Event has passed!";
        long years = (long)(seconds / (365.25 * 24 * 3600));
        seconds -= (long)(years * 365.25 * 24 * 3600);
        long months = (long)(seconds / (30.44 * 24 * 3600));
        seconds -= (long)(months * 30.44 * 24 * 3600);
        long days = seconds / (24 * 3600);
        seconds -= days * (24 * 3600);
        long hours = seconds / 3600;
        seconds -= hours * 3600;
        long minutes = seconds / 60;
        seconds %= 60;
        List<String> parts = new ArrayList<>();
        if (years > 0) parts.add(years + " year" + (years > 1 ? "s" : ""));
        if (months > 0) parts.add(months + " month" + (months > 1 ? "s" : ""));
        if (days > 0) parts.add(days + " day" + (days > 1 ? "s" : ""));
        if (hours > 0) parts.add(hours + " hour" + (hours > 1 ? "s" : ""));
        if (minutes > 0) parts.add(minutes + " minute" + (minutes > 1 ? "s" : ""));
        if (seconds > 0) parts.add(seconds + " second" + (seconds > 1 ? "s" : ""));
        return parts.isEmpty() ? "0 seconds" : String.join(", ", parts);
    }

    static void countdown(Event ev, String formatStyle, boolean live) {
        String[] timeParts = ev.time.split(":");
        int h = Integer.parseInt(timeParts[0]);
        int m = Integer.parseInt(timeParts[1]);
        LocalDate date = parseDate(ev.date);
        if (date == null) return;
        LocalDateTime target = LocalDateTime.of(date, LocalTime.of(h, m));
        LocalDateTime now = LocalDateTime.now();
        long seconds = Duration.between(now, target).getSeconds();
        if (seconds <= 0) {
            System.out.println("⏰ " + ev.name + " is in the past!");
            return;
        }
        if (live) {
            while (seconds > 0) {
                System.out.print("\r⏳ " + ev.name + ": " + formatCountdown(seconds) + "    ");
                try { Thread.sleep(1000); } catch (InterruptedException e) {}
                now = LocalDateTime.now();
                seconds = Duration.between(now, target).getSeconds();
            }
            System.out.println("\r🎉 " + ev.name + " has arrived!    ");
        } else {
            if ("compact".equals(formatStyle)) {
                long totalSec = seconds;
                long days = totalSec / 86400;
                long hours = (totalSec % 86400) / 3600;
                long minutes = (totalSec % 3600) / 60;
                long secs = totalSec % 60;
                System.out.printf("⏳ %s: %02d:%02d:%02d:%02d%n", ev.name, days, hours, minutes, secs);
            } else if ("short".equals(formatStyle)) {
                System.out.println("⏳ " + ev.name + ": " + seconds + " seconds");
            } else {
                System.out.println("⏳ " + ev.name + ": " + formatCountdown(seconds));
            }
        }
    }

    static void listEvents() {
        if (events.isEmpty()) {
            System.out.println("No events saved.");
            return;
        }
        for (Event e : events) {
            String[] timeParts = e.time.split(":");
            int h = Integer.parseInt(timeParts[0]);
            int m = Integer.parseInt(timeParts[1]);
            LocalDate date = parseDate(e.date);
            LocalDateTime target = LocalDateTime.of(date, LocalTime.of(h, m));
            String status = target.isAfter(LocalDateTime.now()) ? "⏳ pending" : "✅ passed";
            System.out.printf("  %s: %s %s - %s%n", e.name, e.date, e.time, status);
        }
    }

    static void interactive() throws Exception {
        Scanner scanner = new Scanner(System.in);
        loadEvents();
        System.out.println("=== Countdown Timer ===");
        while (true) {
            System.out.println("\n1. Set new event");
            System.out.println("2. Show countdown");
            System.out.println("3. Live countdown");
            System.out.println("4. List events");
            System.out.println("5. Delete event");
            System.out.println("6. Save events");
            System.out.println("7. Load events");
            System.out.println("8. Exit");
            System.out.print("Choose: ");
            String choice = scanner.nextLine().trim();
            switch (choice) {
                case "1":
                    System.out.print("Event name: ");
                    String name = scanner.nextLine().trim();
                    if (name.isEmpty()) { System.out.println("Name required."); break; }
                    if (findEvent(name) != null) { System.out.println("Event already exists."); break; }
                    System.out.print("Event date (YYYY-MM-DD): ");
                    String dateStr = scanner.nextLine().trim();
                    if (parseDate(dateStr) == null) { System.out.println("Invalid date."); break; }
                    System.out.print("Event time (HH:MM, optional): ");
                    String timeStr = scanner.nextLine().trim();
                    if (timeStr.isEmpty()) timeStr = "00:00";
                    if (addEvent(name, dateStr, timeStr)) System.out.println("Event saved.");
                    break;
                case "2":
                    System.out.print("Event name: ");
                    name = scanner.nextLine().trim();
                    Event ev = findEvent(name);
                    if (ev == null) { System.out.println("Event not found."); break; }
                    countdown(ev, "full", false);
                    break;
                case "3":
                    System.out.print("Event name: ");
                    name = scanner.nextLine().trim();
                    ev = findEvent(name);
                    if (ev == null) { System.out.println("Event not found."); break; }
                    countdown(ev, "full", true);
                    break;
                case "4": listEvents(); break;
                case "5":
                    System.out.print("Event name to delete: ");
                    name = scanner.nextLine().trim();
                    if (deleteEvent(name)) System.out.println("Event deleted.");
                    else System.out.println("Event not found.");
                    break;
                case "6": saveEvents(); System.out.println("Events saved."); break;
                case "7": loadEvents(); System.out.println("Events loaded."); break;
                case "8": System.out.println("Goodbye!"); scanner.close(); return;
                default: System.out.println("Invalid choice.");
            }
        }
    }

    static void cli(String[] args) {
        Map<String, String> params = new HashMap<>();
        for (int i = 0; i < args.length; i++) {
            if (args[i].startsWith("--")) {
                String key = args[i].substring(2);
                if (i + 1 < args.length && !args[i+1].startsWith("--"))
                    params.put(key, args[++i]);
                else
                    params.put(key, "true");
            }
        }
        loadEvents();
        if (params.containsKey("list")) { listEvents(); return; }
        if (params.containsKey("delete")) {
            if (deleteEvent(params.get("delete"))) System.out.println("Event deleted.");
            else System.out.println("Event not found.");
            return;
        }
        if (params.containsKey("event") && params.containsKey("date")) {
            Event ev = findEvent(params.get("event"));
            if (ev != null) {
                String format = params.getOrDefault("format", "full");
                boolean live = params.containsKey("live");
                countdown(ev, format, live);
            } else System.out.println("Event not found.");
        } else {
            System.out.println("Usage: java Countdown --event NAME --date DATE [--time TIME] [--format full|compact|short] [--live]");
        }
    }

    public static void main(String[] args) throws Exception {
        if (args.length > 0) cli(args);
        else interactive();
    }
}
